### ISSUE TITLE
Test and visualize on custom data

### ISSUE THREAD

**Issue by ttsesm** [zhulf0804/NgeNet#4]
Hi, thanks for sharing your work. Is it easy to give some guidance how to test/evaluate the pre-trained models on two individual given point clouds and/or possibly to custom data?

**Comment by zhulf0804** [comment]
Thanks for your interest.

Yes, i have tried what you referred before.

I'm sorry i can't update the code in time for i am busy in other tasks.

And i'll update it as soon as possible.

Best regards.

**Comment by ttsesm** [comment]
Sure, I was thinking more or less something similar to this https://github.com/yewzijian/RegTR/blob/main/src/demo.py which is a comparative work to yours.

Thanks.

**Comment by zhulf0804** [comment]
Yes, I also think [RegTR](https://github.com/yewzijian/RegTR) is a very excellent work in point cloud registration.

Also, the code style  from [Zi Jian Yew](https://github.com/yewzijian) is worth learning. 

Good luck.

**Comment by ttsesm** [comment]
One more question, going through the paper as well as the source code it is not clear to me how you use the transformation and the correspondence points. If I understand it correctly you use the correspondence points in the circle loss and the transformation matrices in the overlap and saliency losses. Also you compute the overlap and saliency scores on the fly. Is that correct? Can you confirm/elaborate a bit on this?

Btw, what is your opinion for the [Geotransformer](https://github.com/qinzheng93/GeoTransformer) and the [Lepard](https://github.com/rabbityl/lepard) in regards to your approach?

Thanks.

**Comment by zhulf0804** [comment]
In fact, correspondences are obtained through the ground truth transformation matrix, as implemented in function [get_correspondences](https://github.com/zhulf0804/NgeNet/blob/main/utils/o3d.py#L75-L88). The details on how to use the correspondences and transformation can be seen in [loss](https://github.com/zhulf0804/NgeNet/blob/main/losses/loss.py) implementation. But I think it's more important to understand how to define overlap and saliency for supervised learning. 

The second question is very interesting. The following are my **personal** views.

1. Both Lepard and GeoTransformer are CVPR22 Oral, so they should be outstanding works.
2. The highlight of Lepard is to focus on **both rigid and non-rigid registration**. I didn't learn much about the details.
3. GeoTransformer extends CoFiNet (NeurIPS21) with geometric Transformer,  **solving registration in an end-to-end manner**.  RegTR (you referred to above) also proposes an end-to-end Transformer-based registration method. I think the task is challenging and the results are amazing. Nice work.
4. Our work proposes several **universal mechanisms** (ms + voting, GGE module) to **boost registration performance on different registration networks**(FCGF, D3Feat, PREDATOR, etc.). In the paper, we extend PREDATOR-style network with the proposed universal mechanisms as our network.

**Comment by ttsesm** [comment]
Thanks for the elaboration.

**Comment by zhulf0804** [comment]
> Hi, thanks for sharing your work. Is it easy to give some guidance how to test/evaluate the pre-trained models on two individual given point clouds and/or possibly to custom data?

Hi, sorry to the late update. 

A demo for testing on custom point cloud pair is provided [here](https://github.com/zhulf0804/NgeNet#demo), and unseen scene data with the same density (as the pretrained dataset) is supported.

<del> For point cloud with different densities, we find KPConv-based architecture (such as this work) may not generalize well. If KPConv is replaced with MinkowskiEngine, the network generalizes better. <\del> 


Updates:

**Testing data with different voxel size  from the pre-trained dataset is supported now.**
